src/Workspaces/Core/Portable/NamingStyles/Serialization/NamingStylePreferences.cs

```diff
@@ -339,5 +336,36 @@ public override int GetHashCode()
   </NamingRules>
 </NamingPreferencesInfo>
 ";
+
+        private static XElement GetUpgradedSerializationIfNecessary(XElement rootElement)
+        {
+            var serializationVersion = int.Parse(rootElement.Attribute("SerializationVersion").Value);
+
+            if (serializationVersion == 4)
+            {
+                UpgradeSerialization_4To5(rootElement = new XElement(rootElement));
+                serializationVersion = 5;
```

Comment:
one thing i like about this approach is that it's step by step and can be composed.  So, for example, if we added a v6 serialization, we would then just ahve a block below this that says ```if (serializationVersion == 5) { Upgrade_5to6() }"```.  So you write your version->to->next->version code once, and then you just chain more upgrades off of that.  It does mean you potentially do N updates at one point, but that shouldn't be a problem in practice.

Note: should v5 be compatible with V4.  in other words, should we keep ```<SymbolKind>Method</SymoblKind>``` in the xml?  that way old versions are forward compatible with the new version.

Note: should we keep the version at '4'?  otherwise, old clients will read this in and just reset themselves to default.

Reply:
> one thing i like about this approach is that it's step by step and can be composed

Yes I thought about this and that's the purpose why I wrote the code the way I did. I even added a comment for that.

Reply:
@CyrusNajmabadi 
>  in other words, should we keep <SymbolKind>Method</SymoblKind> in the xml
we can't do that. this XML isn't stored anywhere. it's translated into the options object (ignoring unknown properties/elements) which is then serialzied again

Reply:
Interesting.  This may definitely be challenging then if we want a system that can work across multiple versions of VS.  definitely good thing to be discussed at the design meeting.

ideally a design coudl be made that would then allow for this sort of compat *in the future*, potentially codified into some system.  Then we could update our features preemptively to support this 'upgrade/downgrade management' component.  That way, in the future, as we need to augment settings for features, we can do so in a clean way that will work with all vs versions starting at least with the version where we design for this case.

For example, such a component would likely need to be designed such that it simply ignores (but preserves) information it does not understand, versus throwing out that data.  Similar to how protobufs and other formats will allow manipulation of data even with parts they don't understand (for pretty much precisely this sort of forward/backward compat scenario).